### PR TITLE
scripts: build-tools.sh: Make tools individually buildable

### PR DESCRIPTION
**Pull request by crazoes**
More control over which tools should be built, instead of building
everything everytime.

Signed-off-by: Shreeya Patel <shreeya.patel23498@gmail.com>

Related issue: https://github.com/thesofproject/sof/issues/3059

### CODE REVIEW

**Comment by crazoes** [comment]
I think this build fails because of version difference and there's no `-S` option in that version. What version of cmake are we trying to support that I should be using as reference for commands?

**Comment by crazoes** [comment]
CC: @dbaluta @marc-hb

File: scripts/build-tools.sh [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 # fail immediately on any errors
 set -e
+set -x
 
 print_usage()
```
Comment (dbaluta) [inline scripts/build-tools.sh:9]: This commit should be removed right?

File: scripts/build-tools.sh [hunk 2]
```diff
@@ -5,69 +5,106 @@
...
+        fi
+
+        DO_BUILD_CTL=false
         DO_BUILD_FUZZER=false
-        while getopts "tf" OPTION; do
+        DO_BUILD_LOGGER=false
+        DO_BUILD_PROBES=false
+        DO_BUILD_TEST=false
+        DO_BUILD_TOPOLOGIES=false
+        while getopts "cfhlptT" OPTION; do
```
Comment (dbaluta) [inline scripts/build-tools.sh:65]: new line here0x4dc8

File: scripts/build-tools.sh [hunk 3]
```diff
@@ -5,69 +5,106 @@
...
+       [-c] Build ctl
+       [-f] Build fuzzer
+       [-h] Display help
+       [-l] Build logger
+       [-p] Build probes
        [-t] Build test topologies
-       [-f] Build fuzzer"
+       [-T] Build topologies
 EOFUSAGE
 }
```
Comment (dbaluta) [inline scripts/build-tools.sh:22]: Can we compile everything if the script has no parameters?

File: scripts/build-tools.sh [hunk 4]
```diff
@@ -5,69 +5,106 @@
...
+                T) DO_BUILD_TOPOLOGIES=true ;;
+                h) print_usage; exit 1;;
                 *) print_usage; exit 1;;
                 esac
         done
         shift "$(($OPTIND - 1))"
 
-        build_tools
+        generate_make_files
+        if "$DO_BUILD_CTL"
```
Comment (dbaluta) [inline scripts/build-tools.sh:80]: lets call this prepare_build and add a comment that it generates the makefiles

File: scripts/build-tools.sh (hunk 3 above)
Comment (crazoes) [reply scripts/build-tools.sh:22]: yes, it compiles everything if no parameters at https://github.com/thesofproject/sof/pull/3064/files#diff-b3893e3cc0c1ad18c233c8ccb11447ecR55

File: scripts/build-tools.sh (hunk 1 above)
Comment (crazoes) [reply scripts/build-tools.sh:9]: Yes, here because trying to debug the failure, will remove before final review. 

> I think this build fails because of version difference and there's no -S option in that version. What version of cmake are we trying to support that I should be using as reference for commands?

From my previous comment ^^

Comment (dbaluta) [reply scripts/build-tools.sh:9]: @crazoes don't know to answer your questions. Maybe @jajanusz can jump in.

Anyway, why does the build fail now and it worked without this patch? Will make some time later to investigate but this shouldn't happen.

Comment (crazoes) [reply scripts/build-tools.sh:9]: Yes, it works locally. cmake version I wrote this against is 3.17.3

File: scripts/build-tools.sh [hunk 5]
```diff
@@ -5,69 +5,112 @@
...
+prepare_build()
 {
-        cd "$SOF_REPO/tools"
-        rm -rf build_tools
-        mkdir build_tools
-        cd build_tools
+        if ! [ -d "$BUILD_TOOLS_DIR" ]
+        then
+                mkdir "$BUILD_TOOLS_DIR"
+        fi
```
Comment (marc-hb) [inline scripts/build-tools.sh:28]: tl;dr: keep the `rm -rf build_tools`.

This script is rebuilding from scratch and is documented as such
https://thesofproject.github.io/latest/getting_started/build-guide/build-from-scratch.html#step-4-build-topology-and-tools

This change is converting it to incremental builds which is very big change.

In an ideal world, building from scratch is never required. In the real world, incremental builds are frequently broken.

There's also the case of people working on the build system itself. `cmake` tries hard to detect CMakeLists.txt changes and to reconfigure itself but there is only so much it can do.

Anyone working on the code itself (as opposed to the build system) and needing the fastest build time can and should invoke [c]make (or ninja) directly, they don't need and shouldn't use this script. Removing a layer of indirection also allows finer-grained control on the build, think `cmake -DFOO=bar`, `make VERBOSE=1`, etc.

I understand `cmake` and `make` offer a slightly less user-friendly interface but I don't think  that's a huge price to pay if you want incremental builds. One possible way to make  `cmake` and `make` easier could be this, up to you:

```shell
build_tool()
{
    tool=$1
    ( set -x
       make -C ....
    )
}
```
This cheap logging makes it possible to just copy/paste the command for re-use.
It doesn't hurt to see the exact command in CI either.

File: scripts/build-tools.sh [hunk 6]
```diff
@@ -5,69 +5,112 @@
...
+        if ! [ -d "$BUILD_TOOLS_DIR/fuzzer" ]
+        then
+                mkdir "$BUILD_TOOLS_DIR/fuzzer"
+        fi
+        cd "$BUILD_TOOLS_DIR/fuzzer"
+        cmake ../../fuzzer
 }
 
-build_fuzzer()
+build_tool()
```
Comment (marc-hb) [inline scripts/build-tools.sh:41]: As the only thing this does is invoking `make`, can you call it `make_tool()`?

File: scripts/build-tools.sh [hunk 7]
```diff
@@ -5,69 +5,112 @@
...
 {
-        cd "$SOF_REPO/tools/build_tools"
-        mkdir build_fuzzer
-        cd build_fuzzer
-        cmake ../../fuzzer
-        make -j "$NO_PROCESSORS"
+        tool=$1
+        make -C "$BUILD_TOOLS_DIR/$tool" -j "$NO_PROCESSORS"
```
Comment (marc-hb) [inline scripts/build-tools.sh:44]: Please add a comment here that an empty `$tool` is OK and supported.

File: scripts/build-tools.sh [hunk 8]
```diff
@@ -5,69 +5,112 @@
...
+
+        if "$DO_BUILD_LOGGER"
+        then
+                build_tool logger
+        fi
+
+        if "$DO_BUILD_PROBES"
+        then
+                build_tool probes
+        fi
```
Comment (marc-hb) [inline scripts/build-tools.sh:103]: You can use `eval` and a `for` loop to avoid all this repetition:
```shell
DO_BUILD_ctl=true
DO_BUILD_logger=false
for tool in ctl logger probes ... ; do
    if eval '$DO_BUILD_'$tool; then
        build_tool $tool
    fi
done
```
I'm usually not a fan of `eval` as it increases the quoting complexity[*] when most people (including myself) keep getting a single level of quoting wrong. On the other hand, this is a very simple and minimal use of `eval` and it saves a lot of duplication.

The alternative to `eval` is some `tool -> bool` map/associative array. That's much more change though, up to you.

[*] http://rg1-teaching.mpi-inf.mpg.de/unixffb-ss98/quoting-guide.html

File: scripts/build-tools.sh (hunk 1 above)
Comment (marc-hb) [reply scripts/build-tools.sh:8]: From the top of `sof/CMakeLists.txt`:

```
cmake_minimum_required(VERSION 3.10)
```
3.10 is 2.5 years old which is not crazy old. On the other hand there are many repositories and other ways to upgrade cmake easily, so I think we should bump that to 3.13 which has a few useful features like for instance `-B`. We should check the default version in most popular Linux distributions. Off-topic sorry.

File: scripts/build-tools.sh (hunk 5 above)
Comment (crazoes) [reply scripts/build-tools.sh:28]: Yes, docs seem to already cover and recommend the approach you stated. However, if we remove build_tools everytime, having individual switch to build individual tools might not make sense, because if I run `./scripts/build-tools -f`, I get fuzzer, if I do `./scripts/build-tools -p` I will get probes, but the fuzzer will be removed. At any given moment it will only contain tools that were last built by the script.

File: scripts/build-tools.sh (hunk 8 above)
Comment (crazoes) [reply scripts/build-tools.sh]: this is a good trick

Comment (crazoes) [reply scripts/build-tools.sh]:
```
In ./scripts/build-tools.sh line 76:
                l) DO_BUILD_logger=true ;;
                   ^-------------^ SC2034: DO_BUILD_logger appears unused. Verify use (or export if used externally).
```

shellcheck has similar complaint for other variables. Because we are using eval for dynamically using the variable, the static analysis of shellcheck fails to see usage of these variables.

**Comment by aiChaoSONG** [comment]
SOFCI TEST

File: scripts/build-tools.sh (hunk 5 above)
Comment (marc-hb) [reply scripts/build-tools.sh:28]: TODO

File: scripts/build-tools.sh (hunk 8 above)
Comment (marc-hb) [reply scripts/build-tools.sh]: Good to know. It's possible to turn off shellcheck warnings with a magic `# shellcheck ...` comment, or we could use `eval` here too but let's worry about that later when we add shellcheck to `sof` CI (it's already in sof-test). For now I'm just suggesting to run it locally.

File: scripts/build-tools.sh (hunk 5 above)
Comment (marc-hb) [reply scripts/build-tools.sh:28]: I spent a bit more time looking at this. `rm -rf "$BUILD_TOOLS_DIR/$tool"` is not guaranteed to work because the top-level CMakeLists.txt file does not promise that every $tool will be built in a subdirectory name $tool.

> because if I run ./scripts/build-tools -f, I get fuzzer, if I do ./scripts/build-tools -p I will get probes, but the fuzzer will be removed. At any given moment it will only contain tools that were last built by the script.

It depends, one tool/target could depend on another, so you could have more than one.

This made me realize something else: so you apparently want to rebuild one tool without touching the others yet `cmake ..` is re-run at the top-level for ALL the tools!
- `cmake` + `make` for the tools requested
- `cmake` but no make for the tools not requested

That's inconsistent and not every intuitive.

Trying this locally I also found that building everything takes less than 3 seconds on my system.  I also wonder how you tested this because I found the make targets to be called `sof-logger`, `sof-ctl` and `sof-probes`. The  targets `logger` and `ctl` exist but they don't build anything.

I'm wondering what problem(s) this really solves.

Comment (crazoes) [reply scripts/build-tools.sh:28]: > This made me realize something else: so you apparently want to rebuild one tool without touching the others yet `cmake ..` is re-run at the top-level for ALL the tools!
> 
> * `cmake` + `make` for the tools requested
> * `cmake` but no make for the tools not requested

Initially I tried to cmake only the tool asked to build. But when I tried 
```
rm -rf tools/build_tools/
mkdir tools/build_tools/logger
cd tools/build_tools/logger
cmake ../../logger
```
the makefiles were written to `tools/logger`, weird. So I tried `cmake -S tools/logger -B tools/build_tools/logger` which worked and was used initially in my PR. However, that is not supported by the cmake in our CI and the README also suggests CMake > 3.10 hence, only way I found was to cmake the whole tools directory and make it work.

> Trying this locally I also found that building everything takes less than 3 seconds on my system. I also wonder how you tested this because I found the make targets to be called `sof-logger`, `sof-ctl` and `sof-probes`. The targets `logger` and `ctl` exist but they don't build anything.

I didn't test make command inside in the `tools/build_tools` directory. I just tested the script, checked the executables in the build directories and even tested make commands in the individual build directories. Does the targets have to be named `logger`, `ctl`, etc? 

> I'm wondering what problem(s) this really solves.

Only specified tools are built, not other tools.

Comment (dbaluta) [reply scripts/build-tools.sh:28]: @marc-hb we are trying to make the script more flexible so only you compile the things you need. For example, for me it is always a problem when I try to compile the topologies because I always need to export the LD_LIBRARY_PATH to a newer version of alsa-lib so that topology can compile fine.

So, for example I only want to build the fuzzer and not to deal with building the topologies.

**Comment by marc-hb** [comment]
> For example, for me it is always a problem when I try to compile the topologies because I always need to export the LD_LIBRARY_PATH to a newer version of alsa-lib so that topology can compile fine.

Ha, this is not about build speed. I understand better now. I guess you're referring to alsa-lib bug #2543? BTW I have a quick, dirty and lazy hack for this situation: disabling ASRC topologies. I just shared it there.



> I didn't test make command inside in the tools/build_tools directory. I just tested the script, checked the executables in the build directories and even tested make commands in the individual build directories. Does the targets have to be named logger, ctl, etc?

Oops sorry I missed that, I didn't pay enough attention and assumed this PR invoked the targets that can be seen in `make help` So this is another way the current version of this PR violates cmake's "encapsulation": cmake's subdirectories are internal cmake implementation details not supposed to be relied on. They're not part of the "build API". The build API if any is made of the targets printed by `make help` or `ninja -t -targets`


> if I run ./scripts/build-tools -f, I get fuzzer, if I do ./scripts/build-tools -p I will get probes, but the fuzzer will be removed. At any given moment it will only contain tools that were last built by the script.

Thinking about it again, this really doesn't seem like the end of the world. That's just the small price to pay for the lack of build modularity.


  ------

> Initially I tried to cmake only the tool asked to build. [...]

The current structure of the CMakeLists.txt files is simply not designed for that much modularity; that's what I was trying to to highlight.

If the build of some tool(s) is truly independent from others then this should be addressed in the CMakeLists.txt files themselves, not poking at various [c]make internals like this from an additional layer of indirection.

While `make` has some drawbacks, it's most convenient feature is the ability to `make [ -C build ] anything` at the user interface level without writing any line of user interface code - unlike this PR. So here's an alternative to this PR you could consider: instead of adding code for each tool every time, how about a new `./scripts/host-build-tool.sh --only-cmake-you-name-it` option that  runs only cmake and does not run make but prints instead a one-line help/hint like `cd build_tools; make sof-logger sof-fuzzer ...` at the end? It would avoid build issues like #2543 without trying to incompletely duplicate `[c]make` user interface in an additional layer of indirection while relying on internal `[c]make` implementation details.

**Comment by marc-hb** [comment]
I've done some additional CMake research and found this nice entry:
https://stackoverflow.com/questions/33443164/cmake-share-library-with-multiple-executables
The question is very close and the answers very relevant. Other sources generally concur.

Reminder:
- cmake      = build configuration
- make/ninja = actual build

tl;dr:
1. To build a collection of truly independent projects, each with their independent cmake configuration, use `ExternalProject_add()` at the top.
2. To build a collection of librairies/components with a SINGLE, global cmake configuration, use `add_subdirectory()` at the top layer. You can still build one component at a time but not configure one at a time.

Between 1. and 2.  I'm not sure what is the best fit for `sof/tools/`. Maybe it should even be on a per tool basis? But until someone demonstrates `ExternalProject` is better at the top and volunteers to switch to it, we must stick to the approach currently in place. For now this is 2, so partial `cmake` configuration is not supported. It doesn't make sense either to run cmake and regenerate the global configuration while leaving some semi-random collection of former builds behind. Worst case it could even confuse the next build.

**Comment by aiChaoSONG** [comment]
@xiulipan @marc-hb This PR changes the default behavior of build-tools.sh, which will break CI. I suggest you provide `-a` option for all. Sometimes, we do need to build everything, eg, CI

**Comment by aiChaoSONG** [comment]
SOFCI TEST

**Comment by aiChaoSONG** [comment]
Our CI use -t option to build test topology, but with this PR, -t option build nothing.

**Comment by crazoes** [comment]
> Our CI use -t option to build test topology, ctl, etc. This PR will influence CI. @xiulipan

In the CI there were two types of usage 
1. `./scripts/build-tools` - compile all. This is still the case. So don't need to change anything.
2. `./scripts/build-tools -t` - It's my understanding that we don't need to build topology, ctl, etc to run tests `make tests` works regardless. Since the tests are still ran with the new change, didn't change anything here either.

**Comment by aiChaoSONG** [comment]
> > Our CI use -t option to build test topology, ctl, etc. This PR will influence CI. @xiulipan
> 
> In the CI there were two types of usage
> 
> 1. `./scripts/build-tools` - compile all. This is still the case. So don't need to change anything.
> 2. `./scripts/build-tools -t` - It's my understanding that we don't need to build topology, ctl, etc to run tests `make tests` works regardless. Since the tests are still ran with the new change, didn't change anything here either.

Old `-t` option compiles  topology, test topology, clt, logger and probe, but now it only compile test topology.

**Comment by crazoes** [comment]
> If the build of some tool(s) is truly independent from others then this should be addressed in the CMakeLists.txt files themselves, not poking at various [c]make internals like this from an additional layer of indirection.

Doesn't using `make -C tools/build_tools sof-logger` remove the current reliance on the cmake's internal implementation? We can amend the commit to use this instead of going into individual directories and then run make. 

> It doesn't make sense either to run cmake and regenerate the global configuration while leaving some semi-random collection of former builds behind. Worst case it could even confuse the next build.

However, if we add `--only-cmake-you-name-it` option we are basically doing the same thing, just letting user run `make` instead of the script. So, not much difference. In which case we either do both `cmake` and `make` or not do any change at all?

> Between 1. and 2. I'm not sure what is the best fit for `sof/tools/`. Maybe it should even be on a per tool basis? But until someone demonstrates `ExternalProject` is better at the top and volunteers to switch to it, we must stick to the approach currently in place. For now this is 2, so partial `cmake` configuration is not supported. It doesn't make sense either to run cmake and regenerate the global configuration while leaving some semi-random collection of former builds behind. Worst case it could even confuse the next build.

I can give the option no. 1 a shot although this'd be first time me working with cmake files. And if we decide to go with option 2 route then there should be no change at all as I explained earlier in this comment. What do you think?

**Comment by marc-hb** [comment]
> However, if we add --only-cmake-you-name-it option we are basically doing the same thing, just letting user run make instead of the script. So, not much difference

I meant `--only-re-cmake-you-name-it` _and_ `rm -rf build_tools`, sorry for giving too many options and letting you combine them :-) The ruler splitting my comment meant change of topic: the `--only-cmake-you-name-it` suggestion is an alternative to boilerplate code duplicating the existing `[c]make` user interface. Note this option would be... _optional_, by default the script would still reconfigure _and_ rebuild everything in one go which is and has always been its main convenience.

> I can give the option no. 1 (switch to `ExternalProject_add `) a shot although this'd be first time me working with cmake files.

If you think you have enough days and motivation to spend on this then why not but... wait: I'm not a CMake CODEOWNER in this project.
@jajanusz, others: would you be OK with more modular `tools` build starting with something like `sed -i -e s/add_subdirectory/ExternalProject_Add/g sof/tools/CMakeLists.txt` ?

> Doesn't using make -C tools/build_tools sof-logger remove the current reliance on the cmake's internal implementation? 

Yes it does IMHO.

> And if we decide to go with option 2 (keep `add_subdirectory`) route then there should be no change at all as I explained earlier in this comment. 

The very minimal code changes I would like compared to the current version are:
- Invoke make targets at the top level instead of invoking sub directories e.g. `make -C tools/build_tools sof-logger`
- `rm -rf build_tools` every time (due to the `cmake` reconfiguration every time)

Logging (with set -x) the make commands to help users peel off this layer of indirection when needed would be the icing on the cake.

Without any doubt much less work than option 1.

**Comment by crazoes** [comment]
Okay @marc-hb will implement option 2 for now and work on making the tools independent after that's done parallelly with my work on fuzzer.

**Comment by dbaluta** [comment]
> Ha, this is not about build speed. I understand better now. I guess you're referring to alsa-lib bug #2543? BTW I have a quick, dirty and lazy hack for this situation: disabling ASRC topologies. I just shared it there.

Yes, this is not about build speed. It's about flexibility. In our case is that we want to be able to compile just the fuzzer and not the rest of the tools.

**Comment by marc-hb** [comment]
> Yes, this is not about build speed. It's about flexibility.

Flexibility can be desired for a large variety of (good) reasons including smaller/faster builds. I understand the more specific goal is now to avoid unneeded and broken ALSA dependencies.

The flexibility that this PR wants to add is _already_ there. It's just at a lower and slightly less convenient level you don't like, just below this wrapper script.

File: scripts/build-tools.sh [hunk 9]
```diff
@@ -5,69 +5,127 @@
...
+usage: $0 [-c|-f|-h|-l|-p|-t|-T]
+       [-c] Build ctl
+       [-f] Build fuzzer
+       [-h] Display help
+       [-l] Build logger
+       [-p] Build probes
        [-t] Build test topologies
-       [-f] Build fuzzer"
+       [-T] Build topologies
+       [-n] No build, only cmake
```
Comment (crazoes) [inline scripts/build-tools.sh:19]: Couldn't decide on a better single letter option to represent "No build, only cmake". Open to suggestions.

**Comment by marc-hb** [comment]
The `sof-ci/jenkins/pr-fw-build — FW Build Failed:byt_xcc` in https://sof-ci.01.org/sofpr/PR3064/build6424/build/ is "will not fit" issue #3063

File: scripts/build-tools.sh (hunk 9 above)
Comment (marc-hb) [reply scripts/build-tools.sh:19]: Could you change `Build ctl` to `Re-build ctl` etc.?

The `[ ]` brackets are not needed: they mean something is optional which makes sense only in the SYNOPSIS. Look at `man tar` and `tar --help` for a good example.

> Couldn't decide on a better single letter option to represent "No build, only cmake". Open to suggestions.

`-n` is fine, maybe you'll prefer this:
```
usage: $0 [OPTION...]
...
  -C  No build, only CMake re-configuration. Ignored when building specific tool(s).
```
The "ignored" part I added is IMHO important.

File: scripts/build-tools.sh [hunk 10]
```diff
@@ -5,69 +5,123 @@
...
-        DO_BUILD_FUZZER=false
-        while getopts "tf" OPTION; do
+        prepare_build
+
+        if [ $# -eq 0 ]; then
+                if "$DO_BUILD"; then
+                        make_tool
+                        make_fuzzer
+                        exit
+                fi
```
Comment (marc-hb) [inline scripts/build-tools.sh:80]: It's very rare that adding a negation helps but in this particular case I find this easier to understand:
```shell
if ! "$CMAKE_ONLY"; then
```

File: scripts/build-tools.sh [hunk 11]
```diff
@@ -5,69 +5,123 @@
...
         done
         shift "$(($OPTIND - 1))"
 
-        build_tools
-
-        if "$DO_BUILD_TEST"
-        then
-                build_test
+        if ! "$DO_BUILD"; then
+                print_build_info
```
Comment (marc-hb) [inline scripts/build-tools.sh:107]: > It's very rare that adding a negation helps but...

One less negation here - we're even :-)

File: scripts/build-tools.sh (hunk 10 above)
Comment (crazoes) [reply scripts/build-tools.sh:80]: actually removing the conditional altogether. first, `DO_BUILD` is undefined until L89 so `./scripts/build-tools.sh` fails. And when we do specify `-n` the number of options will never be 0, hence, this will never execute, so the additional conditional is unnecessary.

File: scripts/build-tools.sh (hunk 9 above)
Comment (crazoes) [reply scripts/build-tools.sh:19]: Although it is not ignored when building specific tool. If I do `./scripts/build-tools.sh -c -n` it will only generate the configuration and print build instructions.

File: scripts/build-tools.sh [hunk 12]
```diff
@@ -5,69 +5,126 @@
...
-        DO_BUILD_TEST=false
-        DO_BUILD_FUZZER=false
-        while getopts "tf" OPTION; do
+        DO_BUILD_ctl=false
+        DO_BUILD_fuzzer=false
+        DO_BUILD_logger=false
+        DO_BUILD_probes=false
+        DO_BUILD_tests=false
+        DO_BUILD_topologies=false
+        DO_BUILD=false
```
Comment (marc-hb) [inline scripts/build-tools.sh:85]: now CMAKE_ONLY

File: scripts/build-tools.sh [hunk 13]
```diff
@@ -5,69 +5,126 @@
...
+                BUILD_ALL
         SCRIPT_DIR=$(cd "$(dirname "$0")" && pwd)
         SOF_REPO=$(dirname "$SCRIPT_DIR")
+        BUILD_TOOLS_DIR="$SOF_REPO"/tools/build_tools
         : ${NO_PROCESSORS:=$(nproc)}
+        BUILD_ALL=false
+
+        if [ $# -eq 0 ]; then
+                BUILD_ALL=true
+        fi
```
Comment (marc-hb) [inline scripts/build-tools.sh:77]: Can you move this after parsing options, so it's ready for adding a flag that affects all builds in the future?

File: scripts/build-tools.sh (hunk 10 above)
Comment (marc-hb) [reply scripts/build-tools.sh]: Thx I missed one `exit`. Funny how easy it is to get lost in not that much cyclomatic complexity. I think your latest version 1c3dfde303489f reduces it.

File: scripts/build-tools.sh [hunk 14]
```diff
@@ -5,69 +5,126 @@
...
+       -p Rebuild probes
+       -t Rebuild test topologies
+       -T Rebuild topologies
+       -C No build, only CMake re-configuration
 EOFUSAGE
 }
 
-build_tools()
+# generate Makefiles
+prepare_build()
```
Comment (marc-hb) [inline scripts/build-tools.sh:24]: Maybe `reconfigure_build()` to align with CMake language. No big deal.

File: scripts/build-tools.sh (hunk 13 above)
Comment (crazoes) [reply scripts/build-tools.sh:77]: if we move this zero arg test below getopts, we shift everything by then and it will always be 0. Everything gets parsed and no way to know if any arguments were provided(build specific things) or none were provided(build everything).

**Comment by crazoes** [comment]
> Indeed it looks like I got that wrong. So it's the other way round: using `-C` ignores asking for specific tool(s), correct? The current help could be a little bit more explicit but it's OK.

simply put: if `-C ` is specified nothing gets built, ever.

File: scripts/build-tools.sh (hunk 13 above)
Comment (crazoes) [reply scripts/build-tools.sh:77]: Yes, and I was trying to avoid not having to set `BUILD_ALL` with each option provided.

**Comment by crazoes** [comment]
I see [travis test](https://travis-ci.org/github/thesofproject/sof/jobs/702557005) is failing with a timeout for `cnl icl`. Is getting a timeout usual? Can someone restart that stage?

**Comment by marc-hb** [comment]
> I see travis test is failing with a timeout for cnl icl. Is getting a timeout usual?

Yes it is the usual way to stop qemu, please review https://github.com/thesofproject/qemu/pull/44 to make that clearer.

> Can someone restart that stage?

I suggest we merge #3101 `restore qemu errors` first (my bad) so we get some information in case it fails again.

**Comment by aiChaoSONG** [comment]
SOFCI TEST

**Comment by aiChaoSONG** [comment]
SOFCI TEST

**Comment by marc-hb** [comment]
> I suggest we merge #3101 restore qemu errors first (my bad) so we get some information in case it fails again.

#3101 just got merged, @crazoes can you rebase?

**Comment by crazoes** [comment]
@marc-hb everything passes now. And no, I hadn't tried reproducing this locally.

**Comment by aiChaoSONG** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@zrombel any comment on CI failure ? This PR seems unrelated.

**Comment by zrombel** [comment]
@lgirdwood there was a hardware failure over night, PR is already waiting for rebuild.

**Comment by crazoes** [comment]
@lgirdwood just a reminder for this PR to be merged.

**Comment by marc-hb** [comment]
Fixing shellcheck warnings including some typos in variable names in #3365
